### ISSUE TITLE
Adaptive formats are rate limited despite deciphering N

### ISSUE THREAD

**Issue by azihassan** [azihassan/youtube-d#65]
I noticed that downloading with an itag other than 18 or 22 causes a slowdown, even if the N parameter is deciphered. What's strange is that if I abort the slow download and resume it again, the rate limiting no longer applies.

**Comment by azihassan** [comment]
I ran a few tests using the HTML and base.js files from Firefox. The video URL I found in Firefox's network analyser was similar to the one that youtube-d returned (with --output-url or --verbose) :

![image](https://github.com/azihassan/youtube-d/assets/1638227/8f5b1711-4a09-4f40-b732-cf310a20a821)

I suspect that the range argument is what's causing it to be rate limited. Running cURL with and without `-C -` shows that there's a difference in download speeds :

![image](https://github.com/azihassan/youtube-d/assets/1638227/886cbf59-d54f-448f-96b6-1efec4289672)

**Comment by azihassan** [comment]
I noticed that certain videos don't have this behaviour. Here's an example where the download speed is limited to 110-ish kB/s :

![image](https://github.com/azihassan/youtube-d/assets/1638227/b7ad2a16-3a39-428c-9271-4c6e35d59760)

**Comment by azihassan** [comment]
libcurl doesn't send a Range header if [https://curl.se/libcurl/c/CURLOPT_RESUME_FROM.html](CURLOPT_RESUME_FROM) is set to 0, command-line cURL doesn't either. As established in the earlier comments, a missing range parameter or Range header triggers rate-limiting measures for adaptive formats. As a workaround, a `Range: bytes=0-` header needs to be manually added to the request if the destination file doesn't exist.

**Comment by azihassan** [comment]
I did some more tests and noticed that rate limiting isn't applied when requesting small ranges. Beyond a certain range, youtube starts rate limiting adaptive format videos for some reason. In a 200 MB video, the range cut-off was at 11 MBs :

![image](https://github.com/azihassan/youtube-d/assets/1638227/7e75a78f-8ff6-4c5f-a7f6-0c527d96853d)

Not sure if 11 MBs is calculated as a percentage of 200 MB or if it's a hard limit regardless of the video size though.

**Comment by azihassan** [comment]
I changed ParallelDownloader to take this into consideration. Applying it to RegularDownloader turned into a major refactoring, so I gave up on it. Instead I added a ChunkedDownloader that acts like a serial version of ParallelDownloader, meaning that it only downloads one chunk at a time. This will bypass the 11 MB limitation with the added benefit of not spamming the youtube servers with parallel connections. But this comes at the cost of introducing a chunk concatenation step when the download is finished. I'll test drive these changes for a few days to make sure they're stable before merging.
